web/services/collector.go

```diff
@@ -19,11 +22,36 @@ func NewCollectorService(db *gorm.DB, projectorsChannel chan *datapipeline.DataC
 	return &collectorService{db: db, projectorsChannel: projectorsChannel}
 }
 
+// StoreEvent stores the event in the database and sends it to the projectors.
+// If the last event is equal to the current one, it is not stored nor sent.
 func (c *collectorService) StoreEvent(collectedData *datapipeline.DataCollectedEvent) error {
-	if err := c.db.Create(collectedData).Error; err != nil {
-		return err
-	}
-	c.projectorsChannel <- collectedData
+	return c.db.Transaction(func(tx *gorm.DB) error {
+		var event datapipeline.DataCollectedEvent
+		err := tx.
+			Where(&datapipeline.DataCollectedEvent{
+				AgentID:       collectedData.AgentID,
+				DiscoveryType: collectedData.DiscoveryType,
+				Payload:       collectedData.Payload,
+			}).
+			Last(&event).
+			Error
+
+		if errors.Is(gorm.ErrRecordNotFound, err) {
+			err := tx.Create(collectedData).Error
+			if err != nil {
+				return err
+			}
+			c.projectorsChannel <- collectedData
+			return nil
+		}
+
+		if err != nil {
+			return err
+		}
+
+		log.Debugf("Event already exists. Agent ID: %s, DiscoveryType: %s ", collectedData.AgentID, collectedData.DiscoveryType)
+		c.projectorsChannel <- &event
```

Comment:
I'm not quite clear, why are we projecting the last event?

Reply:
We still do not have a mechanism of retrial or queue persistence in the projector system, so if the event wasn't projected for some reason we need to be sure that the projectors will receive the event and try again.
The projectors keep the state of the subscription so if the event was already projected they will just skip it.